src/types/closure.md

```diff
@@ -42,48 +50,153 @@ so that the call to `f` works as if it were:
+
+Moving fields out of references is not allowed. As a result, in the case of move closures, when values accessed through a shared references are moved into the closure body, the compiler will truncate right before a dereference.
+
+```rust
+struct T(String, String);
+
+let mut t = T(String::from("foo"), String::from("bar"));
+let t = &mut t;
+let c = move || t.0.truncate(0); // closure captures `t`
+```
+
+### Raw pointer dereference
+Because it is `unsafe` to dereference a raw pointer, closures will only capture the prefix of a path that runs up to, but not including, the first dereference of a raw pointer.
+
+```rust,
+struct T(String, String);
+
+let t = T(String::from("foo"), String::from("bar"));
+let t = &t as *const T;
+
+let c = || unsafe {
+    println!("{}", (*t).0); // closure captures t
+};
+```
+
+### Reference into unaligned `struct`s
+
+Because it is `unsafe` to hold references to unaligned fields in a structure, closures will only capture the prefix of the path that runs up to, but not including, the first field access into an unaligned structure.
+
+```rust
+#[repr(packed)]
+struct T(String, String);
+
+let t = T(String::from("foo"), String::from("bar"));
+let c = || unsafe {
+    println!("{}", t.0); // closure captures t
+};
+```
+
+
+### `Box` vs other `Deref` implementations
+
+The implementation of the [`Deref`] trait for [`Box`] is treated differently from other `Deref` implementations, as it is considered a special entity.
+
+For example, let us look at examples involving `Rc` and `Box`. The `*rc` is desugared to a call to the trait method `deref` defined on `Rc`, but since `*box` is treated differently by the compiler, the compiler is able to do precise capture on contents of the `Box`.
+
+[`Box`]: ../special-types-and-traits.md#boxt
+[`Deref`]: ../special-types-and-traits.md#deref-and-derefmut
+
+#### Non `move` closure
+
+In a non `move` closure, if the contents of the `Box` are not moved into the closure body, the contents of the `Box` are precisely captured.
+
+```rust
+# use std::rc::Rc;
+
+struct S(i32);
+
+let b = Box::new(S(10));
+let c_box = || {
+    println!("{}", (*b).0); // closure captures `(*b).0`
+};
+
+let r = Rc::new(S(10));
+let c_rc = || {
+    println!("{}", (*r).0); // closure caprures `r`
+};
+```
+
+However, if the contents of the `Box` are moved into the closure, then the box is entirely captured. This is done so the amount of data that needs to be moved into the closure is minimized.
+
+```rust
+struct S(i32);
+
+let b = Box::new(S(10));
+let c_box = || {
+    let x = (*b).0; // closure captures `b`
+};
+```
+
+#### `move` closure
+
+Similarly to moving contents of a `Box` in a non-`move` closure, reading the contents of a `Box` in a `move` closure will capture the `Box` entirely.
+
+```rust
+struct S(i32);
+
+let b = Box::new(S(10));
+let c_box = || {
+    println!("{}", (*b).0); // closure captures `b`
+};
+```
 
 ## Unique immutable borrows in captures
```

Comment:
The example in this section needs to be updated, as it now captures `*x` by mutable borrow, not `x` by unique immutable borrow.